Improvement: Reformat non-embedded to embedded youtube video link

wutschel opened the pull request:
## Description
<!--- Detailed info for reviewers and developers -->
The implemented solution will identify multiple variants of youtube.com and youtu.be links (http, https, www or not), extract the video id and create a fresh FullHD embedded youtube link from this. This allows to show all such links with in-app preview.

## Summary for release notes
<!--- This will be shown in Testflight to end users / testers -->
<!--- Please fill in, as usually PR title isn't clear to ordinary users -->
<!--- If your changes don't modify app files, please add prefix [not app] -->
Improvement: Reformat non-embedded to embedded youtube video link

kambala-decapitator commented on XBMC Remote/ShowInfoViewController.m:
```diff
@@ -1172,21 +1172,37 @@ - (BOOL)isEmbeddedYoutubeLink:(NSString*)urlPath {
+    // Checks for a) "http://" OR "https://", "www" OR not AND "youtube.com/watch?v=" (4 variations)
+    // and b) "http://" OR "https://" AND "youtu.be/watch?v=" (2 variations)
+    BOOL isYoutubeComLink = [urlPath rangeOfString:@"^https?://(www\\.)?youtube\\.com/watch\\?v=" options:NSRegularExpressionSearch].location != NSNotFound;
+    BOOL isYoutuBeLink = [urlPath rangeOfString:@"^https?://youtu.be/watch\\?v=" options:NSRegularExpressionSearch].location != NSNotFound;
```
regex was fine for one-off, but since now we need more advanced logic, let's work with NSURLComponents directly

wutschel replied:
Done

kambala-decapitator commented on XBMC Remote/ShowInfoViewController.m:
```diff
@@ -1172,21 +1172,37 @@ - (BOOL)isEmbeddedYoutubeLink:(NSString*)urlPath {
+        NSArray *queryItems = urlComponents.queryItems;
+        for (NSURLQueryItem *item in queryItems) {
+            if ([item.name isEqualToString:queryItemName]) {
+                embedVideoURL = [NSString stringWithFormat:@"https://www.youtube.com/embed/%@?&vq=hd1080", item.value];
```
please change this function to be "pure" (aka without side effects), i.e. return value and not assign any object ivars/properties inside

wutschel replied:
Making it return `NSString` and call it like `embedVideoURL = [self getVideoURL...]` you mean?

kambala-decapitator replied:
exactly :)

wutschel replied:
Done. Makes this really better readable inside `processTrailerFromString`. Thanks for suggesting.

kambala-decapitator commented on XBMC Remote/ShowInfoViewController.m:
```diff
@@ -1172,21 +1172,37 @@ - (BOOL)isEmbeddedYoutubeLink:(NSString*)urlPath {
 - (void)processTrailerFromString:(NSString*)trailerString {
     embedVideoURL = nil;
     if (trailerString.length > 0) {
         if ([trailerString hasPrefix:@"plugin://plugin.video.youtube"]) {
```
btw this check can also be changed to use NSURLComponents

wutschel replied:
Done, but for `isEmbeddedYoutubeLink` I think I should do this as well, right?

kambala-decapitator commented on XBMC Remote/ShowInfoViewController.m:
```diff
@@ -1168,25 +1168,90 @@ - (void)processCastFromArray:(NSArray*)array {
+    NSURL *url = [NSURL URLWithString:urlPath];
+    if (url) {
+        NSURLComponents *urlComponents = [NSURLComponents componentsWithURL:url resolvingAgainstBaseURL:YES];
+        NSString *path = urlComponents.path;
+        NSString *host = urlComponents.host;
+        NSString *scheme = urlComponents.scheme;
```
I'd avoid these variables

wutschel replied:
If you are not bothered too much I would like to keep it.

kambala-decapitator commented on XBMC Remote/ShowInfoViewController.m:
```diff
@@ -1168,25 +1168,90 @@ - (void)processCastFromArray:(NSArray*)array {
+        NSString *scheme = urlComponents.scheme;
+        if ([scheme isEqualToString:@"http"] ||
+            [scheme isEqualToString:@"https"]) {
+            if ([host isEqualToString:@"www.youtube.com"] ||
+                [host isEqualToString:@"youtube.com"]) {
```
maybe just "contains youtube.com"?

wutschel replied:
This way it does not allow non-working URLs like ww.youtube.com, so I would like to keep this strict.

kambala-decapitator commented on XBMC Remote/ShowInfoViewController.m:
```diff
@@ -1168,25 +1168,90 @@ - (void)processCastFromArray:(NSArray*)array {
+        NSString *scheme = urlComponents.scheme;
+        if ([scheme isEqualToString:@"http"] ||
+            [scheme isEqualToString:@"https"]) {
+            if ([host isEqualToString:@"www.youtube.com"] ||
+                [host isEqualToString:@"youtube.com"] ||
```
these conditions can be extracted to a function to avoid copy-paste

wutschel replied:
But this would make the other check which also includes "youtu.be" harder to read. Let's please keep.

kambala-decapitator replied:
sorry, not sure how it makes it harder: `[self isYoutubeCom:host] || [host isEqualToString:@"youtu.be"]`

kambala-decapitator replied:
so you don't want to change it. Ok, not insisting :)

kambala-decapitator commented on XBMC Remote/ShowInfoViewController.m:
```diff
@@ -1168,25 +1168,90 @@ - (void)processCastFromArray:(NSArray*)array {
+        for (NSURLQueryItem *item in queryItems) {
+            if ([item.name isEqualToString:queryItemName]) {
+                videoURLpath = [NSString stringWithFormat:@"https://www.youtube.com/embed/%@?&vq=hd1080", item.value];
+                break; // We can leave the loop as we found what we were looking for.
```
very useful comment :)

wutschel replied:
Obvious, but there since ages and still a correct statement :)

kambala-decapitator replied:
I'd remove it as it doesn't add any real value

wutschel replied:
Done

kambala-decapitator reviewed (approved):
in general looks good now, but I'd suggest also to do the following:
- use NSURL as ivar instead of NSString: you always create NSURL to pass to NSURLComponents, and in the end you need NSURL to create request anyway
- create NSURL and NSURLComponents outside of the new functions (i.e. do it in `processTrailerFromString`) and pass them to the functions

wutschel commented:
> in general looks good now, but I'd suggest also to do the following:
> 
> * use NSURL as ivar instead of NSString: you always create NSURL to pass to NSURLComponents, and in the end you need NSURL to create request anyway
> * create NSURL and NSURLComponents outside of the new functions (i.e. do it in `processTrailerFromString`) and pass them to the functions

Good point, done.

wutschel commented:
Squashed and rebased as I understood despite the comments it was approved. Just let me know, if you still would like me to do further changes.

kambala-decapitator commented:
> create NSURL **and NSURLComponents** outside of the new functions

seems you missed this point, could you adjust?

wutschel commented:
> seems you missed this point, could you adjust?

Yep, sorry. Was only focusing the `NSURL` part of it. Done now as well, out of consistency for all the link check methods.

wutschel commented:
Squashed and rebased, with latest fixups. More rework than expected for this initially small change, but the implementation is more straight forward now.
